### DISCUSSION TITLE
Make email optional on a Customer

### DISCUSSION THREAD

**Discussion by Psvensso** [vendurehq/vendure#860] [General]
Today Customers have to have an email, this is a deeply integrated part of the application and not easy to override with a plugin. This limits other authentication options that don't rely on emails like Phone OTP or e-signatures. 

I see that the user concept are "identifier type" agnostic, so i could do my own user resolution in my authentication strategy (today i do it via ExternalAuthenticationService methods) but this flow of Customer (basically a lot in CustomerService) would need some rethinking.

Any thoughts?

**Comment by michaelbromley** [comment]
Can you give me an example of something that is problematic as a result of the assumption of an email address? I've not used OTP or other non-email based auth flows myself, so I am probably missing a lot of context here.

**Reply by Psvensso** [reply]
Thanks for taking your time Michael!
 
Iv'e been digging and trying to understand the Vendure codebase but I'm a rookie at it so apologies in advanced if I got some assumptions wrong. 

Background:
In our use-case we don't want to require email from my customers for a few reasons.

We have many customers in our CRM right now that we don't have emails to, many of them are elderly or we just collected their information on farmers markets etc. and simply don't have it. When we do a sale on other sale channels (like via phone) we would like to add the customer to the Webshop's customer lists and even create the order on their behalf. But at this point we don't have an email. But in our case the unique identifier for that customer is the phone number. 
(Creating the order on their behalf of a customer would be a separate plugin and not really part of this discussion).

We can then ask the customer to visit our site and login via OTP using their phone and see their orders there. 

In this use-case the phone number should be treated as unique and basically follow the same security regime as email does today in Vendure. E.g. you cant change it without confirming it and so on. 

But there must be many more similar use-cases so I think a more generic solution would smart. This functionality should be extendible same as the authentication are. Im thinking this could even be an integrated part of authentication. The Native authentication uses a custom field called username and/or email for example. And SMSAuthentication uses phone_no as it's "data bearer". If you wish to update the value (customer wants to change email or phone) it has to go via the AuthStrategy and follow that flow. For example, if it requires validation.

Today the User gets validated but not the method itself. So after a user gets verified he/she can change email and phone but still remain validated. This flow of letting the AuthStrategy be responsible would also take care of that.

**Reply by Psvensso** [reply]
```Typescript
interface Identifyer: {
 strategy: string; //AuthenticationStrategy.name
 data: Record<string, string> //AuthenticationData
 verified: boolean;
}

interface User {
 identifiers: Identifyer[]
 ...
}
```
Is user verified: user.identifiers.any > verified
Get user email: user.identifiers.find.strategy > "native" > data.email
Change user email: NativeAuth.changeIdentifyerValue() > verification flow

And so on. 

Disclamer: 
I do understand this is probably not the way to solve it and that it oversimplified here, I'm just trying to "get some meat" on the discussion.

**Reply by Psvensso** [reply]
The eID flow is very similar as any other OAuth flow. But in eID you often use some sort of national identification number or other SSID instead of email. So I think this boils down to not using the email as the common denominator.

You imagine this by just implementing a Google strategy but instead of storing the email one would store the SSID from Google. If that was supported then I think most other use-cases would fit.

**Reply by Psvensso** [reply]
Had another look as i got home from work and a lot of this is already implemented on the User, it stores the strategy and externalIdentifier. External identifyer for Google is PID i think and UUID for Facebook. Doesn't really matter what. It's the external id. So i think most moving parts are there, its just make email optional on the Customer and to merge the concepts of "identifyer" "authenticationMethods" and "verified" on the User and most of the use cases would be covered..

**Reply by michaelbromley** [reply]
Thanks for all the ideas on this, that's really helpful input.

So here's what I think: right now I can't make any breaking changes to the way auth is handled. The main missing part is having the "verified" status on a per-authentication method basis, rather than a single flag per User.

 For now, you'll have to work with what is there. Hopefully you can work around the current design and still implement what you need. As you do so you'll run into the edges of what is currently possible and it would be useful to document and keep a track of these limits, so that we may improve things in the next major version.

I invite you to open an actual issue rather than a discussion - I think there is enough substance here to warrant an actual feature request. You can do that now or later when you have a more concrete idea of what should be changed.

**Reply by Psvensso** [reply]
Thank you Michael, 
i toally understand this is nothing that's easily fixed, it would be a pretty big breaking change so not expecting anything. Also this is a quite unique use-case anyway. Most stores today just use the email as unique id and dont bother about the other options. So this is a "nice to have" feature i think. But before i continue playing with it and i just wanted to hear some of your thoughts in a discussion format. :)
